caffe2/onnx/onnx_exporter.cc

```diff
@@ -0,0 +1,585 @@
+  nodes.emplace_back(
+      MakeNode("Reshape", {tmp2, const_tensors.back().name()}, {y}));
+
+  return result;
+}
+
+ConvertedResult OnnxExporter::CreateSliceNodes(
+    const caffe2::OperatorDef& def,
+    const std::unordered_map<std::string, caffe2::TensorShape>& shapes) {
+  CAFFE_ENFORCE_EQ(
+      def.input_size(),
+      1,
+      "ONNX Slice operator does not support dynamic slice.");
+  auto result = CommonCaffe2OpToOnnxNodes(def);
+  auto& nodes = result.first;
+  CAFFE_ENFORCE_EQ(nodes.size(), 1);
+  auto& node = nodes.back();
+  const auto& shape = shapes.at(node.input(0));
+
+  std::vector<int64_t> dims;
+  for (auto& attr: *node.mutable_attribute()) {
+    if (attr.name() == "starts") {
+      auto len = attr.ints_size();
+      if (len) {
+        dims.resize(len);
+        std::iota(dims.begin(), dims.end(), 0);
+      }
+    } else if (attr.name() == "ends") {
+      for (int i = 0; i < attr.ints_size(); ++i) {
+        auto end = attr.ints(i);
+        if (end >=0) {
+          continue;
+        }
+        if (end == -1) {
+          end = shape.dims(i);
+        } else {
+          ++end;
+        }
+        attr.set_ints(i, end);
+      }
+    }
+  }
+  if (!dims.empty()) {
+    node.add_attribute()->CopyFrom(MakeAttribute("axes", dims));
+  }
+
+  return result;
+}
+
+ConvertedResult OnnxExporter::CreateReshapeNodes(
+    const caffe2::OperatorDef& def,
+    const std::unordered_map<std::string, caffe2::TensorShape>& shapes) {
+  auto result = CommonCaffe2OpToOnnxNodes(def);
+  auto& nodes = result.first;
+  auto& const_tensors = result.second;
+  CAFFE_ENFORCE_EQ(nodes.size(), 1);
+  auto& node = nodes.back();
+
+  int i = 0;
+  int attr_size = node.attribute_size();
+  for (; i < attr_size; ++i) {
+    const auto& attr = node.attribute(i);
+    if (attr.name() == "shape") {
+      std::vector<int64_t> shape;
+      for (const auto k: attr.ints()) {
+        shape.push_back(k);
+      }
+      const_tensors.emplace_back(CreateOnnxShapeTensor(shape));
+      node.add_input(const_tensors.back().name());
+      break;
+    }
+  }
+  if (i != attr_size) {
+    if (i != attr_size - 1) {
+      node.mutable_attribute()->SwapElements(i, attr_size - 1);
+    }
+    node.mutable_attribute()->RemoveLast();
+  }
+
+  if (node.output_size() == 2) {
+    node.mutable_output()->RemoveLast();
+  }
+
+  return result;
+}
+
+ConvertedResult OnnxExporter::CreateGemmNodes(
+    const caffe2::OperatorDef& def,
+    const std::unordered_map<std::string, caffe2::TensorShape>& shapes) {
+  CAFFE_ENFORCE_EQ(def.input_size(), 3);
+  CAFFE_ENFORCE_GE(def.output_size(), 1);
+  auto x = def.input(0);
+  auto w = def.input(1);
+  const auto& b = def.input(2);
+  const auto& y = def.output(0);
+  const auto& x_shape = shapes.at(x);
+
+  ConvertedResult result;
+  auto& nodes = result.first;
+  auto& const_tensors = result.second;
+  std::unordered_map<std::string, const caffe2::Argument*> args;
```

Comment:
See `ArgumentHelper` in `utils/proto_utils.h`

Reply:
I see. I actually don't want to make an copy of the `Argument`s here.